docs/debugging.rst

```diff
@@ -0,0 +1,249 @@
+It behaves the same as Python-scope ``print``, just being embedded into Taichi kernel.
+
+.. code-block:: python
+
+    x = ti.var(ti.f32, (2, 3))
+    y = 1
+
+    @ti.kernel
+    def inside_taichi_scope():
+        ti.static_print(y)
+        # => 1
+        ti.static_print(x.shape)
+        # => (2, 3)
+        ti.static_print(x.data_type())
+        # => DataType.float32
+        for i in range(4):
+                ti.static_print(i.data_type())
+                # => DataType.int32
+                # will only print once
+
+Unlike ``print``, ``ti.static_print`` will only print the expression once at compile-time, and
+therefore has no runtime cost.
+
+
+Runtime ``assert`` in kernel
+----------------------------
+
+We may use ``assert`` statement in Taichi-scope. When assertion condition failed, a
+``RuntimeError`` will be raised to indicate error.
+
+To make ``assert`` work, first make sure you are using the **CPU backend**.
+For performance reason, ``assert`` only works when ``debug`` mode is on, For example:
+
+.. code-block:: python
+
+    ti.init(arch=ti.cpu, debug=True)
+
+    x = ti.var(ti.f32, 128)
+
+    @ti.kernel
+    def do_sqrt_all():
+        for i in x:
+            assert x[i] >= 0
+            x[i] = ti.sqrt(x)
+
+
+When your debugging work is done, simply set ``debug=False``. Now ``assert`` will be ignored
+therefore no runtime overhead, making your program easy to debug without losing performance.
+
+
+Compile-time ``ti.static_assert``
+---------------------------------
+
+Like ``ti.static_print``, we also provide a static version of ``assert``:
+``ti.static_assert``, it can be useful to make assertion on data type / dimention / shape.
+It works no matter whether ``debug=True`` is specified. When assertion fails, it will
+raise ``AssertionError`` as a Python-scope ``assert`` does.
+
+For example:
+
+.. code-block:: python
+
+    @ti.func
+    def is_odd(x: ti.template()):
+        ti.static_assert(x.data_type() == ti.i32, "is_odd() is only supported for i32")
+        return x % 2 == 1
+
+
+Tips for debugging
+------------------
+
+Debugging a Taichi program can be hard even with the builtin tools above.
+Taichi developers are currently devoting themselves in improving error messages and warnings
+to help user find potential BUGs in their programs.
+
+Here we collected some common BUGs that one might encounter with a Taichi program:
+
+Static typing system
+++++++++++++++++++++
+
+Taichi pretends that it's a dynamical-typed language like Python, but it's actually a
+statically-typed language which will be translated into high performance CPU/GPU instructions.
+
+So the code behavior in Taichi-scope is actually very different from Python-scope!
+
+Type of a variable is simply **determined at its first initialization and never changes later**.
+
+Although static-type provides better performance and simplicity, but may leads to BUGs if
+users not distinguished Taichi-scope from Python-scope, e.g.:
```

Comment:
```suggestion
Although Taichi's static type system provides better performance, it may lead to bugs if
programmers carelessly used the wrong types. For example,
```